nasa/harmony
# Harmony 670 - EKS monitoring and alerting

A PR for CICD code is forthcoming. You may want to test this out there instead if you'd like to see the alert emails get sent to your inbox. (AlertManager will only get deployed in AWS.)

The overall architecture of these changes is:
Prometheus uses its "rules" (PromQL queries) to determine which alerts are "firing" ==> firing alerts are sent via POST request to the Alert Manager ==> Alert Manager routes the alerts to a notification service like AWS SNS

To test this out locally:
- .env changes
  - Set `PROMETHEUS_POD_NOT_HEALTHY_DURATION` and `PROMETHEUS_NODE_NOT_READY_DURATION` to **1m** in your .env file if you want to see the alert sooner than what is set in env-default.
- Alert triggering
  - To test the KubernetesPodNotHealthy alert, before deploying harmony services, delete a service image or two and set the image location(s) in your .env to a nonexistent location (so that when the pod is spun up it can't get to a ready state).
  - To test the KubernetesNodeNotReady alert, tweak the prometheus.yaml where it says `== 0` to say `==1`. I couldn't find a good way of making a node unhealthy, but I think testing the opposite condition should be sufficient to give an idea of what would happen if the node actually went down. (Try pasting variations of `kube_node_status_condition{condition="Ready",status="true"} == $zero_or_one` in the prometheus charting UI if you're not convinced.)

`bin/deploy-prometheus`

`kubectl port-forward {prometheus-pod} 9090:9090 -n monitoring`

Navigate to the prometheus UI 'alerts' tab to see the alerts firing. Note that this is reflecting the changes in prometheus.yaml which is where you configure the alerting rules (not the changes in alert-manager.yaml where the alerts are routed to SNS in deployed environments).

Example email:

<img width="1418" alt="Screen Shot 2022-02-14 at 9 38 46 AM" src="https://user-images.githubusercontent.com/34752045/153884552-d7afdb74-f32f-4310-b741-266d596b8505.png">
